### ISSUE TITLE
Is access tokens (claim based security) supported by Orleans?

### ISSUE THREAD

**Issue by jan-johansson-mr** [dotnet/orleans#3589]
Hi,

Is access tokens (claim based security) supported by Orleans?

**Comment by sergeybykov** [comment]
Orleans leaves security up to application code. For example, access tokens can be validated via interceptors - http://dotnet.github.io/orleans/Documentation/Advanced-Concepts/Interceptors.html.

**Comment by jan-johansson-mr** [comment]
Thanks Sergey!

**Comment by jan-johansson-mr** [comment]
Hi,

One more question, does Orleans support out-of-band data? My though is to flow access token through out-of-band so I do not need to add a parameter (access token argument) to grain methods.

Thanks,
Jan

**Comment by sergeybykov** [comment]
Would `RequestContext` work for that - dotnet.github.io/orleans/Documentation/Advanced-Concepts/Request-Context.html?

**Comment by jan-johansson-mr** [comment]
Yeah :-)

Absolutely, I actually took a peek into the documentation and discovered the request-context in the morning, and realized that this is actually perfect for OOB-data (kind of data that can be pushed in header on some protocols, e.g. SOAP).

When I have time (later this week - by weekend I think), I'll write up an "Authorize" attribute (method, maybe on class level to scope the whole grain), then use request-context to flow a JWT (access token) and see what happens. Very exciting :-)

I was also reading up on Observer pattern, and now I'm on the fence. I was thinking about using RabbitMQ as message bus, but maybe observer with actor would be nice, but then again, maybe not. I've not decided yet...

Thanks!

**Comment by sergeybykov** [comment]
Re: observers. They were originally conceived primarily as a means of sending notifications to clients (frontends). They are best effort and return void because, unlike grains, clients are non-virtual, and cannot be recovered by the runtime in case of a failure.

If you are considering observers for communication *between grains*, then make sure the observer semantics meet your expectations. Otherwise, normal grain calls or streams are likely better options.

**Comment by jan-johansson-mr** [comment]
Thank you for pointing me in the right direction. It works great now with OOB (RequestContext) sprinkled with attributes for AOP, and in my test case JWT for authorization.

The only issue was on the client side, but I opened another issue (wanted some code to fire before sending request to the grain, so I could "automatically" add the token to the RequestContext). I got that issue resolved by help from you guys!
